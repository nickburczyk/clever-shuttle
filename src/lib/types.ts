export type Car = {
    id: number
    brand?: string
    licensePlate?: string
    manufacturer?: string
    operationCity?: string
    status?: CarStatus
    createdAt: Date
    lastUpdatedAt: Date
}

export enum CarStatus {
    'available',
    'in-maintenance',
    'out-of-service',
}