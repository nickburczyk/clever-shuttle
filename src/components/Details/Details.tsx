import { Car } from "../../lib/types"
import styles from "./details.module.css"

interface Props { car: Car }

const Details = ({
    car: {
        id, brand, licensePlate,
        operationCity, status,
        createdAt, lastUpdatedAt,
        manufacturer
    }
}: Props) => (
    <div className={styles.detailsWrapper}>
        <Row label="ID" val={id.toString()} />
        <Row label="Brand" val={brand} />
        <Row label="Manufacturer" val={manufacturer} />
        <Row label="License Plate" val={licensePlate} />
        <Row label="Operation City" val={operationCity} />
        <Row label="Status" val={status?.toString() || ""} />
        <Row label="CreatedAt" val={createdAt.toString()} />
        <Row label="LastUpdatedAt" val={lastUpdatedAt.toString()} />
    </div>
)

export default Details


// DO NOT EXPORT
// Row helper for the above Details Component. 
interface RowProps {
    label: string;
    val?: string;
}

const Row = ({label, val}: RowProps) => (
    <div className={styles.row}>
        <div data-testid={`detail-label`} className={styles.rowLabel}>{label}: </div>
        {!!val && <div data-testid={`detail-value`} className={styles.rowValue}>{val}</div>}
    </div>
)