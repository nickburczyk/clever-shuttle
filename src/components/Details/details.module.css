.detailsWrapper {
    text-align: left;
    margin-bottom: 2rem;
}

.row {
    display: flex;
    padding: .25rem
}

.rowLabel {
    min-width: 150px;
    font-weight: bold;
}
