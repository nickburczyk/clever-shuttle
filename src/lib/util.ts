import { Car } from "./types"

export const API_ADDRESS: string = "https://nckbku0m91.execute-api.eu-central-1.amazonaws.com/cars"

export const fetchCarsList = async():Promise<Car[]> => {
    const url = API_ADDRESS 
    const res = await fetch(url, {
        method: "GET",
    })
    const data = await res.json()
    return data
}