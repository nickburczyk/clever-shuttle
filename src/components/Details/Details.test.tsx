import { render, screen } from '@testing-library/react';
import { create } from 'react-test-renderer'
import Details from './Details'
import { Car, CarStatus } from '../../lib/types';

const car: Car = {
    id: 11111,
    licensePlate: "D-HKY 630",
    manufacturer: "VW",
    brand: "ID Buzz",
    operationCity: "Hannover",
    status: "available",
    createdAt: new Date("2022-01-01"),
    lastUpdatedAt: new Date("2022-09-14"),
}

describe('Home Screen',() => {
    describe("Details Display", () => {
        it ('should match snapshot', () => {
            const tree = create(<Details car={car}/>
            )
            expect(tree.toJSON()).toMatchSnapshot()
        })

        it ('should render labels for all properties on car', () => {
            render(<Details car={car}/>)
            const idLabels = screen.getAllByTestId('detail-label');
            expect(idLabels.length).toEqual(8)
        })

        it ('should render labels for all properties on car', () => {
            render(<Details car={car}/>)
            const idValues = screen.getAllByTestId('detail-value');
            expect(idValues.length).toEqual(8)
        })

        it ('should render all labels, but only render values if val is present', () => {
            const emptyCar:Car={
                id: 123,
                createdAt: new Date("2022-01-01"),
                lastUpdatedAt: new Date("2022-09-14"),
            }
            render(<Details car={emptyCar}/>)
            const idLabels = screen.getAllByTestId('detail-label');
            const idValues = screen.getAllByTestId('detail-value');
            expect(idLabels.length).toEqual(8)
            expect(idValues.length).toEqual(Object.values(emptyCar).length)
        })
    })
})