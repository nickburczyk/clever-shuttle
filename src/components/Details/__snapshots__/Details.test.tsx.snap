// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Screen Details Display should match snapshot 1`] = `
<div
  className="detailsWrapper"
>
  <div
    className="row"
  >
    <div
      className="rowLabel"
      data-testid="detail-label"
    >
      ID
      : 
    </div>
    <div
      className="rowValue"
      data-testid="detail-value"
    >
      11111
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="rowLabel"
      data-testid="detail-label"
    >
      Brand
      : 
    </div>
    <div
      className="rowValue"
      data-testid="detail-value"
    >
      ID Buzz
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="rowLabel"
      data-testid="detail-label"
    >
      Manufacturer
      : 
    </div>
    <div
      className="rowValue"
      data-testid="detail-value"
    >
      VW
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="rowLabel"
      data-testid="detail-label"
    >
      License Plate
      : 
    </div>
    <div
      className="rowValue"
      data-testid="detail-value"
    >
      D-HKY 630
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="rowLabel"
      data-testid="detail-label"
    >
      Operation City
      : 
    </div>
    <div
      className="rowValue"
      data-testid="detail-value"
    >
      Hannover
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="rowLabel"
      data-testid="detail-label"
    >
      Status
      : 
    </div>
    <div
      className="rowValue"
      data-testid="detail-value"
    >
      available
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="rowLabel"
      data-testid="detail-label"
    >
      CreatedAt
      : 
    </div>
    <div
      className="rowValue"
      data-testid="detail-value"
    >
      Sat Jan 01 2022 01:00:00 GMT+0100 (Central European Standard Time)
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="rowLabel"
      data-testid="detail-label"
    >
      LastUpdatedAt
      : 
    </div>
    <div
      className="rowValue"
      data-testid="detail-value"
    >
      Wed Sep 14 2022 02:00:00 GMT+0200 (Central European Summer Time)
    </div>
  </div>
</div>
`;
